_includes/tutorials/multiple-event-types/kafka/markup/dev/make-multi-event-app.adoc

```diff
@@ -0,0 +1,147 @@
+
+Now lets take a look at the Avro consumer.
+[source, java]
+.KafkaConsumer for multiple events in Avro
+----
+consumerRecords.forEach(consumerRec -> {
+  SpecificRecord avroRecord = consumerRec.value(); <1>
+  if (avroRecord instanceof Purchase) {    <2>
+      Purchase purchase = (Purchase) avroRecord;  <3>
+      eventTracker.add(purchase.getItem());
+  } else if (avroRecord instanceof PageView) {
+      PageView pageView = (PageView) avroRecord;
+      eventTracker.add(pageView.getUrl());
+
+// details left out for clarity
+----
+
+<1> Getting the record
+<2> Doing an `instanceof` check to detrmine the type
+<3> Casting to the appropriate concrete type
+
+With the Avro consumer you'll need to use the Java `instanceof` operator to determine concrete type for the record.  Notice that here you're using the `SpecificRecord` interface which every Avro generated object implements.  Once you find the correct concrete type you cast the record to that type and extract the required inforation.
+
+Before you go on to create the application we should mention quickly about the deserialization configurations you need to set
+
+[source, java]
+.Configurations needed by the Consumer for both Avro and Protobuf
+----
+ avroConsumeConfigs.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, KafkaAvroDeserializer.class); <1>
+ avroConsumeConfigs.put(KafkaAvroDeserializerConfig.SPECIFIC_AVRO_READER_CONFIG, true); <2>
+
+ protoConsumeConfigs.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, KafkaProtobufDeserializer.class); <3>
+ protoConsumeConfigs.put(KafkaProtobufDeserializerConfig.SPECIFIC_PROTOBUF_VALUE_TYPE, CustomerEventProto.CustomerEvent.class); <4>
+----
+
+<1> Specifying to use the Avro deserializer for the Avro consumer
+<2> Setting the Avro deserializer to use the specific reader
+<3> Configurations for the Protobuf consumer to use the Protobuf deserializer
+<4> Setting the specific class type for the Protobuf deserializer
+
+It should come as no surprise that you need to set the deserializer class to either `KafkaAvroDeserializer` and `KafkaProtobufDeserializer` for the Avro and Protobuf consumers.  But when working with multiple types, you still need to set the configuration for a specific type.  For Protobuf it's straight forward, setting the specific type to the outer class makes sense since the proto deserialization process knows how to handle the embedded types due to the schema.  With Avro though even with the `union` schema you'll still need to specify to set the `SPECIFIC_AVRO_READER_CONFIG` to `true` to get the concete types.
+
+
+Now go ahead and create the `src/main/java/io/confluent/developer/MultiEventProduceConsumeApp.java` file:
+
++++++
+<pre class="snippet"><code class="java">{% include_raw tutorials/multiple-event-types/kafka/code/src/main/java/io/confluent/developer/MultiEventProduceConsumeApp.java %}</code></pre>
```

Comment:
For consideration: should there be 2 different apps, one for Avro and one for Protobuf?  It's a bit confusing that the same app handles both, so it's harder to grok and it's not directly usable.  If it were 2 separate apps, then the more discrete code might be easier to grok.

Reply:
Ack, agreed